### PR TITLE
Feature/automation upload

### PR DESCRIPTION
**Pull request by lpusok**

### CODE REVIEW

File: bitriseio/bitriseio.go [hunk 1]
```diff
@@ -1,49 +1,59 @@
...
-	bitriseClient.SetSelectedAppSlug(selectedAppSlug)
+	return client, nil
+}
 
+// UploadCodesigningFiles ...
+func UploadCodesigningFiles(client *bitrise.Client, certificates models.Certificates, profiles []models.ProvisioningProfile) (bool, bool, error) {
 	var provProfilesUploaded bool
-	if !certsOnly {
-		provProfilesUploaded, err = uploadExportedProvProfiles(bitriseClient, profiles, outputDirPath)
+	if len(profiles) != 0 {
```
Comment (godrei) [inline bitriseio/bitriseio.go]: what about the `certsOnly` mode?

File: cmd/scan.go [hunk 2]
```diff
@@ -18,14 +18,12 @@ and export the require code signing files.`,
 }
 
 var (
-	isAllowExport    = false
```
Comment (godrei) [inline cmd/scan.go]: why did you remove this flag?

File: codesign/export.go [hunk 3]
```diff
@@ -1,21 +1,106 @@
...
+		return false, false, err
+	}
+
+	provisioningProfiles, err := collectAndExportProvisioningProfiles(profiles)
+	if err != nil {
+		return false, false, err
+	}
+
+	var client *bitrise.Client
+	if uploadConfig.isValid() {
```
Comment (godrei) [inline codesign/export.go]: what if the user provided only the access token by an accident? i would log a warning that you need to use both flags together

File: cmd/scan.go (hunk 2 above)
Comment (lpusok) [reply cmd/scan.go]: It was not used anywhere.

File: bitriseio/bitriseio.go (hunk 1 above)
Comment (lpusok) [reply bitriseio/bitriseio.go]: In that case len(profiles) should equal 0.

File: cmd/xcode.go [hunk 4]
```diff
@@ -40,6 +44,10 @@ func init() {
 	xcodeCmd.Flags().StringVar(&paramXcodeProjectFilePath, "file", "", "Xcode Project/Workspace file path")
 	xcodeCmd.Flags().StringVar(&paramXcodeScheme, "scheme", "", "Xcode Scheme")
 	xcodeCmd.Flags().StringVar(&paramXcodebuildSDK, "xcodebuild-sdk", "", "xcodebuild -sdk param. If a value is specified for this flag it'll be passed to xcodebuild as the value of the -sdk flag. For more info about the values please see xcodebuild's -sdk flag docs. Example value: iphoneos")
+	// Flags used to automatically upload artifacts
+	xcodeCmd.Flags().BoolVar(&isWriteFiles, "write-files", true, "Set wether to export artifacts to a local directory.")
```
Comment (godrei) [inline cmd/xcode.go]: pls also add this to the xamarin and uitest scanner, may these could be root level flags

File: cmd/xcode.go [hunk 5]
```diff
@@ -40,6 +44,10 @@ func init() {
...
+	xcodeCmd.Flags().StringVar(&personalAccessToken, "auth-token", "", "Personal access token. In case app-slug parameter is also provided, will automatically upload artifacts to bitrise.io.")
```
Comment (godrei) [inline cmd/xcode.go]: i would mention that these flags are working only together:

`Personal access token. Requires app-slug flag to be defined as well and will case the codesigning file to be uploaded automatically`

File: cmd/xcode.go [hunk 6]
```diff
@@ -62,9 +70,13 @@ func initExportOutputDir() (string, error) {
 }
 
 func scanXcodeProject(cmd *cobra.Command, args []string) error {
-	absExportOutputDirPath, err := initExportOutputDir()
-	if err != nil {
-		return fmt.Errorf("failed to prepare Export directory: %s", err)
+	absExportOutputDirPath := ""
```
Comment (godrei) [inline cmd/xcode.go]: NOTE: var absExportOutputDirPath string, would explicitly show that we do not care / use the default value of the var

File: cmd/xcode.go [hunk 7]
```diff
@@ -151,7 +162,18 @@ func scanXcodeProject(cmd *cobra.Command, args []string) error {
 		return ArchiveError{toolXcode, err.Error()}
 	}
 
-	certsUploaded, provProfilesUploaded, err := codesigndoc.ExportCodesignFiles(archivePath, absExportOutputDirPath, certificatesOnly, isAskForPassword)
+	certificatesToExport, profilesToExport, err := codesigndoc.CollectCodesignFiles(archivePath, certificatesOnly)
+	if err != nil {
+		return err
+	}
+	certsUploaded, provProfilesUploaded, err := codesign.UploadAndWriteCodesignFiles(certificatesToExport,
```
Comment (godrei) [inline cmd/xcode.go]: pls leave a comment that certificatesOnly will lead to 0 elements certsUploaded, that's why no need to pass it further

File: codesign/export.go [hunk 8]
```diff
@@ -1,21 +1,106 @@
...
+		}
+		fmt.Println()
+		if shouldUpload, err := goinp.AskForBoolFromReader(uploadConfirmMsg, os.Stdin); err != nil {
+			return false, false, err
+		} else if shouldUpload {
+			client, err = bitriseio.GetInteractiveConfigClient()
+		}
+	}
+
+	provProfilesUploaded := (len(profiles) == 0)
```
Comment (godrei) [inline codesign/export.go]: why is it a good default value? in this case actually nothing is uploaded, is this the previous behavior?

File: codesign/export.go [hunk 9]
```diff
@@ -1,21 +1,106 @@
...
+	certsUploaded := (len(certificates) == 0)
+	if client != nil {
+		certsUploaded, provProfilesUploaded, err = bitriseio.UploadCodesigningFiles(client, identities, provisioningProfiles)
+		if err != nil {
+			return false, false, err
+		}
+	}
+
+	if strings.TrimSpace(outputDirPath) != "" {
```
Comment (godrei) [inline codesign/export.go]: is `outputDirPath` always expanded? current dir can be passed by an empty string, would not it case issue?

File: gows.yml [hunk 10]
```diff
@@ -1 +1 @@
-package_name: github.com/bitrise-tools/codesigndoc
+package_name: github.com/bitrise-io/codesigndoc
```
Comment (godrei) [inline gows.yml]: could be removed as we do not use `gows` this days

File: bitriseio/bitrise/profile.go [hunk 11]
```diff
@@ -215,12 +215,10 @@ func (client *Client) RegisterProvisioningProfile(provisioningProfSize int64, ex
 }
 
 // UploadProvisioningProfile ...
-func (client *Client) UploadProvisioningProfile(uploadURL string, uploadFileName string, outputDirPath string, exportFileName string) error {
-	log.Printf("Upload %s to Bitrise...", exportFileName)
+func (client *Client) UploadProvisioningProfile(uploadURL string, uploadFileName string, content io.Reader) error {
```
Comment (lszucs) [inline bitriseio/bitrise/profile.go]: Same comment and previously: `uploadFileName` is used only in the logging statement.

File: bitriseio/bitrise/certificate.go [hunk 12]
```diff
@@ -217,12 +217,10 @@ func (client *Client) RegisterIdentity(certificateSize int64) (RegisterIdentityD
 }
 
 // UploadIdentity ...
-func (client *Client) UploadIdentity(uploadURL string, uploadFileName string, outputDirPath string, exportFileName string) error {
+func (client *Client) UploadIdentity(uploadURL string, uploadFileName string, exportFileName string, content io.Reader) error {
 	log.Printf("Upload %s to Bitrise...", exportFileName)
```
Comment (lszucs) [inline bitriseio/bitrise/certificate.go]: `exportFileName` is used only in the log statement, however, the logging could be done on the caller side.

In that case, the parameter is not needed and the function is cleaner in terms of the log output it produces

File: codesign/export.go (hunk 9 above)
Comment (lpusok) [reply codesign/export.go]: It is at the moment, in initExportOutputDir()

File: codesign/export.go (hunk 8 above)
Comment (lpusok) [reply codesign/export.go]: It is the previous behaviour, the possible reasoning behind it: if no cert are found, do not write the message to manually upload files in printFinished().

File: cmd/scan.go [hunk 13]
```diff
@@ -3,29 +3,77 @@ package cmd
...
+	personalAccessToken string
+	appSlug             string
 )
 
 func init() {
 	RootCmd.AddCommand(scanCmd)
 	scanCmd.PersistentFlags().BoolVar(&isAskForPassword, "ask-pass", false, "Ask for .p12 password, instead of using an empty password")
 	scanCmd.PersistentFlags().BoolVar(&certificatesOnly, "certs-only", false, "Collect Certificates (Identities) only")
+	var writeFilesRaw string
+	scanCmd.PersistentFlags().StringVar(&writeFilesRaw, writeFilesFlag, "", "Set wether to export artifacts to a local directory.")
```
Comment (godrei) [inline cmd/scan.go]: pls include the default value and the available options in the flag description

File: cmd/xamarin.go [hunk 14]
```diff
@@ -159,35 +161,48 @@ func scanXamarinProject(cmd *cobra.Command, args []string) error {
...
+		fmt.Println(stringutil.LastNLines(logOutput, 15))
+		fmt.Println()
+		if isLogFileWritten {
 			log.Warnf("Please check the logfile (%s) to see what caused the error", logOutputFilePath)
 			log.Warnf(`and make sure that you can "Archive for Publishing" this project from Xamarin!`)
 			fmt.Println()
 			log.Infof("Open the project: %s", xamarinCmd.SolutionFilePath)
 			log.Infof(`And do "Archive for Publishing", after selecting the Configuration+Platform: %s|%s`, xamarinCmd.Configuration, xamarinCmd.Platform)
 			fmt.Println()
 		}
```
Comment (godrei) [inline cmd/xamarin.go]: else if writeFiles == codesign.WriteFilesFallback ?

File: cmd/scan.go [hunk 15]
```diff
@@ -3,29 +3,77 @@ package cmd
...
+			{
+				writeFiles = codesign.WriteFilesAlways
+			}
+		case "always":
+			{
+				writeFiles = codesign.WriteFilesAlways
+			}
+		case "fallback":
+			{
+				writeFiles = codesign.WriteFilesFallback
```
Comment (godrei) [inline cmd/scan.go]: it seems it is never used
